kqueue.c

```diff
@@ -278,47 +262,27 @@ kq_dispatch(struct event_base *base, struct timeval *tv)
 		ts_p = &ts;
 	}
 
-	/* Build "changes" from "base->changes" */
-	EVUTIL_ASSERT(kqop->changes);
+	/* Build the changelist of kevent() from "base->changelist" */
+	EVUTIL_ASSERT(kqop->events);
 	n_changes = kq_build_changes_list(&base->changelist, kqop);
 	if (n_changes < 0)
 		return -1;
 
 	event_changelist_remove_all_(&base->changelist, base);
 
-	/* steal the changes array in case some broken code tries to call
-	 * dispatch twice at once. */
-	changes = kqop->changes;
-	kqop->changes = NULL;
-
-	/* Make sure that 'events' is at least as long as the list of changes:
-	 * otherwise errors in the changes can get reported as a -1 return
-	 * value from kevent() rather than as EV_ERROR events in the events
-	 * array.
-	 *
-	 * (We could instead handle -1 return values from kevent() by
-	 * retrying with a smaller changes array or a larger events array,
-	 * but this approach seems less risky for now.)
-	 */
-	if (kqop->events_size < n_changes) {
-		int new_size = kqop->events_size;
-		do {
-			new_size *= 2;
-		} while (new_size < n_changes);
-
-		kq_grow_events(kqop, new_size);
-		events = kqop->events;
-	}
+	/* Offload the events array before calling kevent() to register and
+	 * retrieve events in case some broken code tries to call dispatch
+	 * while there is already one on the fly. */
+	events = kqop->events;
+	kqop->events = NULL;
 
 	EVBASE_RELEASE_LOCK(base, th_base_lock);
-
-	res = kevent(kqop->kq, changes, n_changes,
-	    events, kqop->events_size, ts_p);
-
+	res = kevent(kqop->kq, events, n_changes, events, kqop->events_size, ts_p);
```

Comment:
Does passing `changelist==eventlist` safe for `kevent`? Please describe why

Reply:
> The same array may be used for the changelist and eventlist.

cited from the man pages of freebsd/macos/dragonflybsd/netbsd/openbsd